lib/msf/core/post/windows/services.rb

```diff
@@ -197,32 +226,72 @@ def each_service(&block)
   # @todo Rewrite to allow operating on a remote host
   #
   def service_list
+    return meterpreter_service_list if session.type == 'meterpreter'
+
+    serviceskey = 'HKLM\\SYSTEM\\CurrentControlSet\\Services'
+    services = []
+    keys = registry_enumkeys(serviceskey)
+    keys.each do |s|
+      service_type = registry_getvaldata("#{serviceskey}\\#{s}", 'Type').to_s
+      service_type = (service_type.starts_with?('0x') ? service_type.to_i(16) : service_type.to_i)
+
+      next unless [
+        SERVICE_WIN32_OWN_PROCESS,
+        SERVICE_WIN32_OWN_PROCESS_INTERACTIVE,
+        SERVICE_WIN32_SHARE_PROCESS,
+        SERVICE_WIN32_SHARE_PROCESS_INTERACTIVE
+      ].include?(service_type)
+
+      services << {:name => s}
+    end
+
+    services
+  end
+
+  def meterpreter_service_list
     if load_extapi
```

Comment:
This seems to do the same thing as the fallback in the new `meterpreter_service_list` without the threaded processing, I there a reason to duplicate?

I suspect it may have been related to this comment:
> Fix a threading issue for Meterpreter sessions when ExtAPI was not loaded, causing inconsistent results with some services missing sometimes.

Another thought, if removing the threading solves some interaction, do you believe the threading add enough value (speed or otherwise) to keep the duplication?

```suggestion

    if session.type == 'meterpreter' && load_extapi
```

Reply:
> This seems to do the same thing as the fallback in the new `meterpreter_service_list` without the threaded processing, I there a reason to duplicate?

The threading already exists before this PR. It does the same thing, but threaded.

> Another thought, if removing the threading solves some interaction, do you believe the threading add enough value (speed or otherwise) to keep the duplication?

Removing the threading solves nothing now that the threading has been fixed. It is slightly faster.

Reply:
Given these details having only one pattern to maintain would be preferred, would you consolidate this, please?

Reply:
> Given these details having only one pattern to maintain would be preferred, would you consolidate this, please?

I don't understand what you want. Your concern is that these lines appear more than once? Or is it that you don't want the threading any more?

```ruby
            service_type = registry_getvaldata("#{serviceskey}\\#{s}", 'Type').to_i

            next unless [
              SERVICE_WIN32_OWN_PROCESS,
              SERVICE_WIN32_OWN_PROCESS_INTERACTIVE,
              SERVICE_WIN32_SHARE_PROCESS,
              SERVICE_WIN32_SHARE_PROCESS_INTERACTIVE
            ].include?(service_type)

            services << {:name => s}
```

Reply:
I was unclear sorry. I would like to see one implementation of registry based enum, threaded or not. Looking closely:

https://github.com/rapid7/metasploit-framework/pull/16928/files#diff-1a1385c73713e0a75c802ab772468dfcab711e5811e8042e09e26f86de6dae82R193-R205

calls `yield sk` in the loop

https://github.com/rapid7/metasploit-framework/pull/16928/files#diff-1a1385c73713e0a75c802ab772468dfcab711e5811e8042e09e26f86de6dae82R231-R243

https://github.com/rapid7/metasploit-framework/pull/16928/files#diff-1a1385c73713e0a75c802ab772468dfcab711e5811e8042e09e26f86de6dae82R260-R282

call `services << {:name => s}` in the loop

It looks to be possible to consolidate on the same code although there is some difference in how the loop values are returned and two have the `service_type.starts_with?('0x')` ternary check. 

Not a required change just feedback that seems viable.

Reply:
> I would like to see one implementation of registry based enum, threaded or not.

I think the threading is worth keeping. Threading is unlikely to work well on non-Meterpreter sessions.

Reply:
Understood, that is a good reason in my book. Marking this as resolved.